context-propagation/src/main/java/io/smallrye/mutiny/context/BaseContextPropagationInterceptor.java

```diff
@@ -0,0 +1,376 @@
+package io.smallrye.mutiny.context;
+
+import java.util.concurrent.Executor;
+import java.util.function.*;
+
+import org.eclipse.microprofile.context.ThreadContext;
+
+import io.smallrye.context.impl.Contextualized;
+import io.smallrye.mutiny.infrastructure.CallbackDecorator;
+import io.smallrye.mutiny.tuples.Functions;
+
+public abstract class BaseContextPropagationInterceptor implements CallbackDecorator {
+
+    /**
+     * Gets the Context Propagation ThreadContext. External
+     * implementations may implement this method.
+     *
+     * @return the ThreadContext
+     */
+    protected abstract ThreadContext getThreadContext();
+
+    @Override
+    public <T> Supplier<T> decorate(Supplier<T> supplier) {
+        if (supplier instanceof Contextualized) {
+            return supplier;
+        }
+        return getThreadContext().contextualSupplier(supplier);
+    }
+
+    @Override
+    public <T> Consumer<T> decorate(Consumer<T> consumer) {
+        if (consumer instanceof Contextualized) {
+            return consumer;
+        }
+        return getThreadContext().contextualConsumer(consumer);
+    }
+
+    @Override
+    public LongConsumer decorate(LongConsumer consumer) {
+        if (consumer instanceof Contextualized) {
+            return consumer;
+        }
+        Consumer<Long> cons = getThreadContext().contextualConsumer(consumer::accept);
```

Comment:
I've added a way for `contextualXXX` to return the parameter unchanged if there are zero contexts to propagate, which can happen if you tell CP that you don't care, using config or the new scoped annotation to override this, if Mutiny starts using it to obtain the `ThreadContext`.

Well, if you do start using this you will depend on SR-CP anyway, but I can add a method `boolean SmallRyeThreadContext.isEmpty()` to allow you to skip calling `contextualConsumer` for nothing in this case, because while it will return its parameter, you will not know and keep wrapping it with `ContextualLongConsumer`.

Reply:
+1, as we depend on SR-CP, let's use have is isEmpty `method`.

Reply:
OK, will add.

Reply:
just started using it :-D